Kernel/Devices/AudioDevice.cpp

```diff
@@ -0,0 +1,454 @@
+    dbgln_if(AUDIODEVICE_DEBUG, "AudioDevice::start_request for card {}", request.card());
+    if (card.state != Card::State::Prepared && card.state != Card::State::Running) {
+        dbgln("AudioDevice::start_request: Card {} not prepared or not in running state", request.card());
+        request.complete(AsyncDeviceRequest::Failure);
+        return;
+    }
+    card.current_request = &request;
+    card.request_buffer_offset = 0;
+    if (card.type == Audio::CardType::Playback) {
+        while (do_transfer_request_buffer_playback(card))
+            ;
+    } else {
+        // TODO: Implement
+        VERIFY_NOT_REACHED();
+    }
+}
+
+bool AudioDevice::do_transfer_request_buffer_playback(Card& card)
+{
+    // This function copies up to one period of data until the next period boundary
+    VERIFY(card.buffer_read_offset <= card.buffer_write_offset);
+    VERIFY(card.buffer_read_offset % card.bytes_per_period == 0);
+    size_t available_space = card.bytes_all_periods - (card.buffer_write_offset - card.buffer_read_offset);
+    if (available_space == 0)
+        return false;
+
+    size_t period_offset = card.buffer_write_offset % card.bytes_per_period;
+    size_t bytes_to_copy = card.bytes_per_period - period_offset;
+    VERIFY(bytes_to_copy <= available_space);
+    size_t request_bytes_remaining = card.current_request->buffer_size() - card.request_buffer_offset;
+    bool write_to_dma = false;
+    bool period_full = false;
+    if (bytes_to_copy > request_bytes_remaining)
+        bytes_to_copy = request_bytes_remaining;
+    else
+        period_full = true;
+
+    u8* dest_ptr;
+    if (card.dma_periods > 0 && card.buffer_write_offset - card.buffer_read_offset < card.bytes_per_period * card.dma_periods) {
+        // Either we're too close to consuming those periods or we haven't triggered playback,
+        // write directly to the dma buffer
+        dest_ptr = playback_current_dma_period(card, true) + period_offset;
+        write_to_dma = true;
+        dbgln("AudioDevice::do_transfer_request_buffer_playback writing {} bytes to dma at offset {}", bytes_to_copy, period_offset);
+    } else {
+        dest_ptr = card.buffer_region->vaddr().as_ptr();
+        dest_ptr += card.buffer_write_offset % card.bytes_all_periods;
+        dbgln("AudioDevice::do_transfer_request_buffer_playback writing {} bytes to buffer at offset {}", bytes_to_copy, period_offset);
+    }
+
+    if (!card.current_request->buffer().read(dest_ptr, bytes_to_copy)) {
+        complete_current_request(card, AsyncDeviceRequest::RequestResult::MemoryFault);
+        return false;
+    }
+    
+    card.buffer_write_offset += bytes_to_copy;
+    card.request_buffer_offset += bytes_to_copy;
+
+    if (card.state == Card::State::Prepared && period_full && (card.current.periods_trigger == 0 || card.buffer_write_offset / card.bytes_per_period == card.current.periods_trigger)) {
+        if (!period_full && card.current.periods_trigger == 0) {
```

Comment:
The inner `if` will never be entered since `period_full` is always `true` per the outer `if.